### DISCUSSION TITLE
posts/Go/go_mysql/

### DISCUSSION THREAD

**Discussion by Q1mi** [Q1mi/BlogComments#350] [Announcements]
https://www.liwenzhou.com/posts/Go/go_mysql/ 

 李文周的Blog Go语言 MySQL 数据库 sqlx db database

**Comment by jihaojie** [comment]
博客写的牛逼！

**Comment by ShouCaoHe** [comment]
作者那个Rows调用了next函数后也不用defer关闭把，next在返回false的时候会自动调用Rows的Close方法。。

**Comment by Q1mi** [comment]
> @ShouCaoHe
> 作者那个Rows调用了next函数后也不用defer关闭把，next在返回false的时候会自动调用Rows的Close方法。。

防的就是那些没有调用Next的情况

**Comment by zt102545** [comment]
现在GO流行的ORM是什么？xorm吗？

**Comment by xuelang201201** [comment]
@Q1mi 大佬，练习题咋做的呀？视频里面没有答案。

**Comment by xiaolongbao321** [comment]
#### 写的很粗糙
```go
// 处理错误的方法

// internal error
type Er struct {
	Code    int
	Message string
}

func (e *Er) Error() string {
	return e.Message
}

// api error
type Err struct {
	Code    int
	Message string
	Err     error
}

func (err *Err) Error() string {
	return fmt.Sprintf("Err - code: %d, message: %s, error: %s", err.Code, err.Message, err.Err)
}

func New(er *Er, err error) *Err {
	return &Err{Code: er.Code, Message: er.Message, Err: err}
}

// main.go
type userRegisterInfo struct {
	Name string	`json:"name"`
	Age int	`json:"age"`
	Passwd string	`json:"passwd"`
	Gander string	`json:"gander"`
	ConformPasswd string `json:"conform_passwd"`
}

type userLoginInfo struct {
	Name string `json:"name"`
	Passwd string `json:"passwd"`
}

var db *sql.DB
const dsn = "root:123456@tcp(127.0.0.1:3306)/sql_test"

func indexHandler(w http.ResponseWriter, r *http.Request){
	fmt.Fprintln(w, "Hello A用户！")
}

func registerHandler(w http.ResponseWriter, r *http.Request){
	defer r.Body.Close()
	var answer = `{"status": "ok"}`
	userInfo := &userRegisterInfo{}
	// 1. 请求类型是application/x-www-form-urlencoded时解析form数据
	//r.ParseForm()
	//fmt.Println(r.PostForm)
	//fmt.Println(r.PostForm.Get("name"), r.PostForm.Get("passwd"), r.PostForm.Get("age"), r.PostForm.Get("gander"))
	// 2. 请求类型是application/json时从r.Body读取数据
	b, err := ioutil.ReadAll(r.Body)
	if err != nil{
		fmt.Printf("read request.Body failed, err:%v\n", err)
		return
	}
	fmt.Println(string(b))
	err = json.Unmarshal(b, userInfo)
	if err != nil{
		fmt.Printf("解析参数时出错：%v\n", err.Error())
		return
	}
	err = Register(userInfo)
	if err != nil{
		err = errnum.New(&errnum.Er{40001, "用户注册时出错"}, err)
		answer = `{"status": "error"}`
		w.Write([]byte(answer + err.Error()))
		return
	} else {
		w.Write([]byte(answer))
		return
	}
}

func loginHandler(w http.ResponseWriter, r *http.Request){
	defer r.Body.Close()
	var answer = `{"status": "ok", "data": "登录成功！"}`
	userInfo := &userLoginInfo{}

	b, err := ioutil.ReadAll(r.Body)
	if err != nil{
		fmt.Printf("read request.Body failed, err:%v\n", err)
		return
	}
	fmt.Println(string(b))
	err = json.Unmarshal(b, userInfo)
	if err != nil{
		fmt.Printf("解析参数时出错：%v\n", err.Error())
		return
	}

	err = Login(userInfo)
	if err != nil{
		err = errnum.New(&errnum.Er{40002, "用户登录时出错"}, err)
		answer = `{"status": "error"}`
		w.Write([]byte(answer + err.Error()))
		return
	} else {
		w.Write([]byte(answer))
		return
	}


	w.Write([]byte(answer))
}

func Register(userInfo *userRegisterInfo) (err error){
	// 校验用户输入的两次密码是否一致
	if userInfo.Passwd != userInfo.ConformPasswd{
		err = errnum.New(&errnum.Er{50001, "请保证两次输入密码的一致性"}, nil)
		return err
	}

	// 连接数据库
	db, err = sql.Open("mysql", dsn)
	if err != nil{
		err = errnum.New(&errnum.Er{50002, "打开数据库连接失败！"}, err)
		return err
	}
	// 测试连接数据库
	err = db.Ping()
	if err != nil{
		err = errnum.New(&errnum.Er{50003, "连接数据库失败"}, err)
		return err
	}

	// 查询当前用户是否已经存在
	existsSqlStr := "select id from users where name = ?"
	stmt, err := db.Prepare(existsSqlStr)
	if err != nil{
		err = errnum.New(&errnum.Er{50004, "预处理查询SQL失败"}, err)
		return err
	}

	defer stmt.Close()
	rows, err := stmt.Query(userInfo.Name)
	if err != nil{
		err = errnum.New(&errnum.Er{50005, "查询失败"}, err)
		return err
	}

	if rows.Next(){
		err = errnum.New(&errnum.Er{50006, "注册用户已存在"}, nil)
		return err
	}

	// 如果注册用户不存在，则插入用户信息
	insertSqlStr := "insert into users (name, passwd, age, gander) values (?, ?, ?, ?)"
	stmt, err = db.Prepare(insertSqlStr)
	if err != nil{
		err = errnum.New(&errnum.Er{50007, "预处理插入SQL失败"}, err)
		return err
	}
	_, err = stmt.Exec(userInfo.Name, userInfo.Passwd, userInfo.Age, userInfo.Gander)
	if err != nil{
		err = errnum.New(&errnum.Er{50008, "插入用户信息失败"}, err)
		return err
	}

	return nil
}

func Login(userInfo * userLoginInfo) (err error){
	var u userLoginInfo
	// 创建数据库连接
	db, err = sql.Open("mysql", dsn)
	if err != nil{
		err = errnum.New(&errnum.Er{50002, "打开数据库连接失败！"}, err)
		return err
	}

	err = db.Ping()
	if err != nil{
		err = errnum.New(&errnum.Er{50003, "连接数据库失败"}, err)
		return err
	}

	// 校验查询用户是否存在；如果存在，则校验用户的密码是否正确
	searchSql := "select name, passwd from users where name = ?"

	stmt, err := db.Prepare(searchSql)
	if err != nil{
		err = errnum.New(&errnum.Er{50004, "预处理查询SQL失败"}, err)
		return err
	}
	defer stmt.Close()
	rows, err := stmt.Query(userInfo.Name)
	if err != nil{
		err = errnum.New(&errnum.Er{50005, "查询失败"}, err)
		return err
	}
	if !rows.Next(){
		err = errnum.New(&errnum.Er{51001, "登录用户不存在，请先注册"}, nil)
		return err
	}
	err = rows.Scan(&u.Name, &u.Passwd)
	if err != nil{
		err = errnum.New(&errnum.Er{51002, "解析单行数据出错"}, err)
		return err
	}

	if u.Passwd != userInfo.Passwd{
		err = errnum.New(&errnum.Er{51003, "用户输入的密码有误"}, err)
		return err
	}

	return nil
}

func main() {
	http.HandleFunc("/index", indexHandler)
	http.HandleFunc("/register", registerHandler)
	http.HandleFunc("/login", loginHandler)
	err := http.ListenAndServe(":9999", nil)
	if err != nil{
		fmt.Printf("http server failed, err:%v\n", err)
		return
	}
}
```

**Comment by callmePicacho** [comment]
> @xiaotuotuo321
> #### 写的很粗糙
> ```go
> // 处理错误的方法
> 
> // internal error
> type Er struct {
> 	Code    int
> 	Message string
> }
> 
> func (e *Er) Error() string {
> 	return e.Message
> }
> 
> // api error
> type Err struct {
> 	Code    int
> 	Message string
> 	Err     error
> }
> 
> func (err *Err) Error() string {
> 	return fmt.Sprintf("Err - code: %d, message: %s, error: %s", err.Code, err.Message, err.Err)
> }
> 
> func New(er *Er, err error) *Err {
> 	return &Err{Code: er.Code, Message: er.Message, Err: err}
> }
> 
> // main.go
> type userRegisterInfo struct {
> 	Name string	`json:"name"`
> 	Age int	`json:"age"`
> 	Passwd string	`json:"passwd"`
> 	Gander string	`json:"gander"`
> 	ConformPasswd string `json:"conform_passwd"`
> }
> 
> type userLoginInfo struct {
> 	Name string `json:"name"`
> 	Passwd string `json:"passwd"`
> }
> 
> var db *sql.DB
> const dsn = "root:123456@tcp(127.0.0.1:3306)/sql_test"
> 
> func indexHandler(w http.ResponseWriter, r *http.Request){
> 	fmt.Fprintln(w, "Hello A用户！")
> }
> 
> func registerHandler(w http.ResponseWriter, r *http.Request){
> 	defer r.Body.Close()
> 	var answer = `{"status": "ok"}`
> 	userInfo := &userRegisterInfo{}
> 	// 1. 请求类型是application/x-www-form-urlencoded时解析form数据
> 	//r.ParseForm()
> 	//fmt.Println(r.PostForm)
> 	//fmt.Println(r.PostForm.Get("name"), r.PostForm.Get("passwd"), r.PostForm.Get("age"), r.PostForm.Get("gander"))
> 	// 2. 请求类型是application/json时从r.Body读取数据
> 	b, err := ioutil.ReadAll(r.Body)
> 	if err != nil{
> 		fmt.Printf("read request.Body failed, err:%v\n", err)
> 		return
> 	}
> 	fmt.Println(string(b))
> 	err = json.Unmarshal(b, userInfo)
> 	if err != nil{
> 		fmt.Printf("解析参数时出错：%v\n", err.Error())
> 		return
> 	}
> 	err = Register(userInfo)
> 	if err != nil{
> 		err = errnum.New(&errnum.Er{40001, "用户注册时出错"}, err)
> 		answer = `{"status": "error"}`
> 		w.Write([]byte(answer + err.Error()))
> 		return
> 	} else {
> 		w.Write([]byte(answer))
> 		return
> 	}
> }
> 
> func loginHandler(w http.ResponseWriter, r *http.Request){
> 	defer r.Body.Close()
> 	var answer = `{"status": "ok", "data": "登录成功！"}`
> 	userInfo := &userLoginInfo{}
> 
> 	b, err := ioutil.ReadAll(r.Body)
> 	if err != nil{
> 		fmt.Printf("read request.Body failed, err:%v\n", err)
> 		return
> 	}
> 	fmt.Println(string(b))
> 	err = json.Unmarshal(b, userInfo)
> 	if err != nil{
> 		fmt.Printf("解析参数时出错：%v\n", err.Error())
> 		return
> 	}
> 
> 	err = Login(userInfo)
> 	if err != nil{
> 		err = errnum.New(&errnum.Er{40002, "用户登录时出错"}, err)
> 		answer = `{"status": "error"}`
> 		w.Write([]byte(answer + err.Error()))
> 		return
> 	} else {
> 		w.Write([]byte(answer))
> 		return
> 	}
> 
> 
> 	w.Write([]byte(answer))
> }
> 
> func Register(userInfo *userRegisterInfo) (err error){
> 	// 校验用户输入的两次密码是否一致
> 	if userInfo.Passwd != userInfo.ConformPasswd{
> 		err = errnum.New(&errnum.Er{50001, "请保证两次输入密码的一致性"}, nil)
> 		return err
> 	}
> 
> 	// 连接数据库
> 	db, err = sql.Open("mysql", dsn)
> 	if err != nil{
> 		err = errnum.New(&errnum.Er{50002, "打开数据库连接失败！"}, err)
> 		return err
> 	}
> 	// 测试连接数据库
> 	err = db.Ping()
> 	if err != nil{
> 		err = errnum.New(&errnum.Er{50003, "连接数据库失败"}, err)
> 		return err
> 	}
> 
> 	// 查询当前用户是否已经存在
> 	existsSqlStr := "select id from users where name = ?"
> 	stmt, err := db.Prepare(existsSqlStr)
> 	if err != nil{
> 		err = errnum.New(&errnum.Er{50004, "预处理查询SQL失败"}, err)
> 		return err
> 	}
> 
> 	defer stmt.Close()
> 	rows, err := stmt.Query(userInfo.Name)
> 	if err != nil{
> 		err = errnum.New(&errnum.Er{50005, "查询失败"}, err)
> 		return err
> 	}
> 
> 	if rows.Next(){
> 		err = errnum.New(&errnum.Er{50006, "注册用户已存在"}, nil)
> 		return err
> 	}
> 
> 	// 如果注册用户不存在，则插入用户信息
> 	insertSqlStr := "insert into users (name, passwd, age, gander) values (?, ?, ?, ?)"
> 	stmt, err = db.Prepare(insertSqlStr)
> 	if err != nil{
> 		err = errnum.New(&errnum.Er{50007, "预处理插入SQL失败"}, err)
> 		return err
> 	}
> 	_, err = stmt.Exec(userInfo.Name, userInfo.Passwd, userInfo.Age, userInfo.Gander)
> 	if err != nil{
> 		err = errnum.New(&errnum.Er{50008, "插入用户信息失败"}, err)
> 		return err
> 	}
> 
> 	return nil
> }
> 
> func Login(userInfo * userLoginInfo) (err error){
> 	var u userLoginInfo
> 	// 创建数据库连接
> 	db, err = sql.Open("mysql", dsn)
> 	if err != nil{
> 		err = errnum.New(&errnum.Er{50002, "打开数据库连接失败！"}, err)
> 		return err
> 	}
> 
> 	err = db.Ping()
> 	if err != nil{
> 		err = errnum.New(&errnum.Er{50003, "连接数据库失败"}, err)
> 		return err
> 	}
> 
> 	// 校验查询用户是否存在；如果存在，则校验用户的密码是否正确
> 	searchSql := "select name, passwd from users where name = ?"
> 
> 	stmt, err := db.Prepare(searchSql)
> 	if err != nil{
> 		err = errnum.New(&errnum.Er{50004, "预处理查询SQL失败"}, err)
> 		return err
> 	}
> 	defer stmt.Close()
> 	rows, err := stmt.Query(userInfo.Name)
> 	if err != nil{
> 		err = errnum.New(&errnum.Er{50005, "查询失败"}, err)
> 		return err
> 	}
> 	if !rows.Next(){
> 		err = errnum.New(&errnum.Er{51001, "登录用户不存在，请先注册"}, nil)
> 		return err
> 	}
> 	err = rows.Scan(&u.Name, &u.Passwd)
> 	if err != nil{
> 		err = errnum.New(&errnum.Er{51002, "解析单行数据出错"}, err)
> 		return err
> 	}
> 
> 	if u.Passwd != userInfo.Passwd{
> 		err = errnum.New(&errnum.Er{51003, "用户输入的密码有误"}, err)
> 		return err
> 	}
> 
> 	return nil
> }
> 
> func main() {
> 	http.HandleFunc("/index", indexHandler)
> 	http.HandleFunc("/register", registerHandler)
> 	http.HandleFunc("/login", loginHandler)
> 	err := http.ListenAndServe(":9999", nil)
> 	if err != nil{
> 		fmt.Printf("http server failed, err:%v\n", err)
> 		return
> 	}
> }
> ```

感谢老哥，花了一天时间跟着走通了原生前后端的注册登录

**Comment by yaomogui** [comment]
感谢大佬的分享

**Comment by biuaxia** [comment]
交作业了～

```golang
package main

import (
	"database/sql"
	"errors"
	"fmt"
	"log"
	"net/http"
	"strings"
	"time"

	_ "github.com/go-sql-driver/mysql"
)

var db *sql.DB

func initMySQL() {
	var err error
	dsn := "q1mi:scxyB33sNT5xX3Ky@tcp(129.28.192.85:3306)/q1mi?charset=utf8mb4&parseTime=True"
	db, err = sql.Open("mysql", dsn)
	if err != nil {
		log.Fatalln("cannot open db, err: ")
		return
	}

	err = db.Ping()
	if err != nil {
		log.Fatalln("connect to db failed, err: ", err)
		return
	}
	db.SetConnMaxLifetime(time.Second * 10)
	db.SetMaxOpenConns(200)
	db.SetMaxIdleConns(10)

	log.Println("connect to db success")
}

func signUp(username, password string) (err error) {
	var stmt *sql.Stmt
	sqlStr := "select password from user where `name` = ?"
	stmt, err = db.Prepare(sqlStr)
	if err != nil {
		return errors.New(fmt.Sprintf("prepare failed, err:%v", err))
	}
	defer stmt.Close()

	rows, err := stmt.Query(username)
	if err != nil {
		return errors.New(fmt.Sprintf("query failed, err:%v", err))
	}
	defer rows.Close()

	dbPassword := ""
	if rows.Next() {
		err := rows.Scan(&dbPassword)
		if err != nil {
			return errors.New(fmt.Sprintf("scan failed, err:%v", err))
		}
	}

	if dbPassword != "" {
		return errors.New(fmt.Sprintf("username:%s exist", username))
	}

	sqlStr = "insert into user(`name`,`password`) values(?,?)"
	stmt, err = db.Prepare(sqlStr)
	if err != nil {
		return errors.New(fmt.Sprintf("prepare failed, err:%v", err))
	}
	defer stmt.Close()

	result, err := stmt.Exec(username, password)
	if err != nil {
		return errors.New(fmt.Sprintf("insert failed, err:%v", err))
	}

	rowsAffected, err := result.RowsAffected()
	if rowsAffected != 1 {
		return errors.New(fmt.Sprintf("insert failed, rowsAffected:%v", rowsAffected))
	}

	return nil
}

func signUpHandler(w http.ResponseWriter, r *http.Request) {
	if err := r.ParseForm(); err != nil {
		_, _ = w.Write([]byte(`{"code": 0, "msg": "form data failed"}`))
		return
	}

	username := strings.TrimSpace(r.Form.Get("username"))
	password := strings.TrimSpace(r.Form.Get("password"))
	if 0 >= len(username) || 0 >= len(password) {
		_, _ = w.Write([]byte(`{"code": 0, "msg": "username or password not empty"}`))
		return
	}

	if err := signUp(username, password); err != nil {
		_, _ = w.Write([]byte(fmt.Sprintf(`{"code": 0, "msg": "%v"}`, err)))
		return
	}

	_, _ = w.Write([]byte(`{"code": 1, "msg": "ok"}`))
	return
}

func signIn(username, password string) (err error) {
	var stmt *sql.Stmt
	sqlStr := "select password from user where `name` = ?"
	stmt, err = db.Prepare(sqlStr)
	if err != nil {
		return errors.New(fmt.Sprintf("prepare failed, err:%v", err))
	}
	defer stmt.Close()

	rows, err := stmt.Query(username)
	if err != nil {
		return errors.New(fmt.Sprintf("query failed, err:%v", err))
	}
	defer rows.Close()

	dbPassword := ""
	if rows.Next() {
		err := rows.Scan(&dbPassword)
		if err != nil {
			return errors.New(fmt.Sprintf("scan failed, err:%v", err))
		}
	}

	if dbPassword == "" {
		return errors.New(fmt.Sprintf("username:%s not exist", username))
	}

	if dbPassword != password {
		return errors.New(fmt.Sprintf("username:%s password:%s not equals", username, password))
	}
	return nil
}

func signInHandler(w http.ResponseWriter, r *http.Request) {
	if err := r.ParseForm(); err != nil {
		_, _ = w.Write([]byte(`{"code": 0, "msg": "form data failed"}`))
		return
	}

	username := strings.TrimSpace(r.Form.Get("username"))
	password := strings.TrimSpace(r.Form.Get("password"))
	if 0 >= len(username) || 0 >= len(password) {
		_, _ = w.Write([]byte(`{"code": 0, "msg": "username or password not empty"}`))
		return
	}

	if err := signIn(username, password); err != nil {
		_, _ = w.Write([]byte(fmt.Sprintf(`{"code": 0, "msg": "%v"}`, err)))
		return
	}

	_, _ = w.Write([]byte(`{"code": 1, "msg": "ok"}`))
	return
}

func main() {
	// 1. 初始化数据库连接
	initMySQL()

	// 2. 设置http路由
	http.HandleFunc("/signIn", signInHandler)
	http.HandleFunc("/signUp", signUpHandler)

	err := http.ListenAndServe(":8080", nil)
	if err != nil {
		fmt.Printf("http server failed, err:%v", err)
		return
	}
}
```
